### PR TITLE
chore(url): add jsdoc typescript checking

### PR DESCRIPTION
**Pull request by dsifford**
## Description

This is the first of many PRs that seek to introduce JSDoc TypeScript type checking to packages. I chose this package as the first one because it's one of the easier ones to do and should be a good starting point for reviewers to familiarize themselves.

I'll make inline comments to some of the changes that I think are most notable. Otherwise, this should be fairly straightforward.

One final thing to note: I decided to add in a `tsconfig.json` file alongside the existing [jsconfig.json](https://github.com/WordPress/gutenberg/blob/master/jsconfig.json) file because the configurations as of right this moment are uniquely different and serve different purposes. Once all the packages have been updated with better JSDoc, we can either keep both of these files OR delete the `tsconfig.json` and merge in the changes into the `jsconfig.json`, whatever you all prefer.

## How has this been tested?

All tests still passing.

## Types of changes

Code quality / Documentation.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/url/src/index.js
Comment (dsifford) [inline packages/url/src/index.js]: This was added to more closely represent the possible return types of `getQueryArg`.

It appears that it is currently possible for the response to be an object, containing keys whose values are either `string` or `string[]` or simply a `string` or `string[]` by itself. 

This definition accounts for that. There needs to be two separate definitions because they are both cyclic and recursive.

File: packages/url/src/index.js
Comment (dsifford) [inline packages/url/src/index.js]: Note `QueryArgParsed` is used here.

File: tsconfig.json
Comment (dsifford) [inline tsconfig.json]: The `include` array will be changed with each new package that is added. We're skipping tests because almost universally throughout the project, incorrect partial types are fed to functions to test a specific case which is not allowed as far as TypeScript is concerned. Fixing all of those instances would require a pretty significant amount of work and there's no real benefit to doing it.

**Review by dsifford — COMMENTED** [review]
Final note: a bunch of the checks are failing for whatever reason. Those are unrelated to the changes I made here as far as I can tell.

File: packages/url/src/index.js
Comment (youknowriad) [inline packages/url/src/index.js]: The default value seems like a duplication for me, can't it be infered?

**Review by youknowriad — COMMENTED** [review]
How can we "use" these checks? Can we add a CI job to check the types or add it to the lint job?

File: packages/url/src/index.js
Comment (dsifford) [reply packages/url/src/index.js]: This is mostly important for documentation generation. Otherwise, I'm mainly just following the WordPress JSDoc standards posted [here](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions).

To answer your question: Yes, the TypeScript compiler can infer this.

**Comment by dsifford** [comment]
> How can we "use" these checks? Can we add a CI job to check the types or add it to the lint job?

Yes, but I was under the impression that it was preferred that I hold off on doing that until the rest of the packages are finished. If not, I can add it into the CI flow no problem.

To run the TypeScript compiler locally yourself, it's as easy as running `npx tsc` from anywhere in the project.

-----

Speaking of checks, what's going on with the checks for this PR? Lots of actions are failing.

**Comment by dsifford** [comment]
Ping @youknowriad -- See comments above. Is this good to merge?

**Comment by youknowriad** [comment]
> Yes, but I was under the impression that it was preferred that I hold off on doing that until the rest of the packages are finished. If not, I can add it into the CI flow no problem.

Can we try this, fees like we can add this pretty easily to the lint job right?

**Comment by dsifford** [comment]
@youknowriad Done.

File: package.json
Comment (youknowriad) [inline package.json]: Should we install the typescript dev dependency? I think travis is failing because of it?

Comment (dsifford) [reply package.json]: Whoops, sorry about that. On it now.

**Comment by youknowriad** [comment]
There's one last failing, I think you should probably just rebase the branch for it to pass.

**Comment by dsifford** [comment]
Re: The failing step --- I think this is unrelated to this PR specifically since that same check was failing from all PRs since (IIRC) 2 days before this PR until (at least) 2 days ago.

Do you still want me to rebase anyway? If so, I'll do it on Monday. I'll be away all weekend.

**Comment by gziolo** [comment]
> Do you still want me to rebase anyway? If so, I'll do it on Monday. I'll be away all weekend.

Yes, it would be great to rebase. Riad is off for the next two weeks, but I can help to land it :)

File: package.json
Comment (gziolo) [reply package.json]: Should we include this command in `lint-staged` section of this file to have it executed as a pre-commit hook?

File: packages/url/README.md
Comment (sirreal) [inline packages/url/README.md]: I find the `[string]` syntax strange. Where does it come from?

I assume it means "optional parameter with a default value", but that only seems to vaguely align with jsdoc, where `[square brackets]` are used on the parameter or property names to indicate they're optional, not around the type.

To add to the confusion, in TypeScript `[ string ]` would mean a one-tuple of strings, or an array with a single `string` member, e.g. `[ "foo" ]` but never `[]` or `[ "foo", "bar" ]`.

File: packages/url/src/index.js
Comment (sirreal) [inline packages/url/src/index.js]: Regarding the return types here in general, it's been bothering me a bit that some are `SomeType | void` and another is `SomeType | undefined`…

It appears that `type void = undefined | null` or with `strictNullChecks` it's `type void = undefined` (our case). However, when I think of functions that return `void`, I tend to think of side-effects. Using `void` in a union with another type to indicate we may or may not get a value seems like something better expressed by `SomeType | undefined`.

As far as I can tell and for our purposes, `void` and `undefined` are type aliases, so this is very much a stylistic preference, but is this something you considered and do you have thoughts on it?

**Review by sirreal — COMMENTED** [review]
This is great, I'm excited to start leveraging all your work around TypeScript. I've left some notes and questions on the PR.

> One final thing to note: I decided to add in a `tsconfig.json` file alongside the existing [jsconfig.json](https://github.com/WordPress/gutenberg/blob/master/jsconfig.json) file because the configurations as of right this moment are uniquely different and serve different purposes. Once all the packages have been updated with better JSDoc, we can either keep both of these files OR delete the `tsconfig.json` and merge in the changes into the `jsconfig.json`, whatever you all prefer.

This note leaves me with additional questions, can you elaborate a bit more? Particularly, it's unclear yo me how they can be "uniquely different and serve different purposes", but in the future one of them would become redundant.

File: tsconfig.json
Comment (sirreal) [reply tsconfig.json]: I tried disabling this and the only error was missing types for `qs` module. Do you expect we'll start pulling in type declarations as `devDependencies`?

I'm on the fence because it seems much easier to relax things when you're too strict than to become more strict when you've been too relaxed 🙂 

Have you already investigated parts of the code where strictness here becomes overly cumbersome and could you provide an example?

File: packages/url/src/index.js
Comment (sirreal) [reply packages/url/src/index.js]: I noticed that in the TypeScript playground, a `return` statement —even empty `return;`— is required to return `undefined` instead of `void`. Viewed from this perspective and accustomed to JavaScript's implicit `undefined` return, that seems like a strange decision 🤷‍♂ 

It's tempting to go add empty returns to the functions that are returning `string | void`, but TypeScript should be supporting us and not the other way around.

These are fine to leave, thanks for your thoughts 👍

File: package.json
Comment (gziolo) [reply package.json]: We can skip it for now if it produces issues, no worries. As long as Travis is going to catch those violations, we are good 👍

File: tsconfig.json
Comment (sirreal) [reply tsconfig.json]: > I'm nervous that we'd lose people who bought into this crazy idea if I told them we'd have to depend on more external types.

Good point 👍

**Review by sirreal — APPROVED** [review]
This looks great to me, thanks @dsifford 🎉 

[There's one improvement I'd like to see for types in the generated documentation](https://github.com/WordPress/gutenberg/pull/17014/files#r317982877), but that's not introduced in this PR and can be addressed later.

**Comment by gziolo** [comment]
> [There's one improvement I'd like to see for types in the generated documentation](https://github.com/WordPress/gutenberg/pull/17014/files#r317982877), but that's not introduced in this PR and can be addressed later.

It's your call whether you want to include it in this PR or do you plan to work on a follow-up. You should be able to merge yourself if you pick the first option. I trust @sirreal more than myself in the context of TypeScript 😅

File: packages/url/src/index.js
Comment (gziolo) [reply packages/url/src/index.js]: My only worry is that we will have to update all occurrences where the return value is optional. It might be quite a huge number of lines. For what it's worth, it's at list very explicit.

File: tsconfig.json
Comment (gziolo) [reply tsconfig.json]: Yes, let's skip tests for now and revisit our approach once we have all packages enabled here 👍

File: tsconfig.json
Comment (gziolo) [inline tsconfig.json]: At some point, you will have to include:
- `"./packages/**/benchmark/**",`

Just saying, there are a few packages which use them :)

**Comment by gziolo** [comment]
> We can try to get something together in this PR if you'd prefer that, but it might require a bit of work and some design input from the team to get done properly. The changes would have to occur in the package that generates the markdown documentation.

I don't think I can help this week, so it's totally fine to open an issue and merge it in the current shape.

**Comment by gziolo** [comment]
I filed a follow-up #17271 to keep track of issues raised in https://github.com/WordPress/gutenberg/pull/17014#pullrequestreview-280249955 by @sirreal.

**Pull request merged**
